coordinator/src/orderbook/trading.rs

```diff
@@ -133,39 +156,75 @@ impl Trading {
-                &mut connection,
+            let opposite_direction_orders = orders::all_by_direction_and_type(
+                &mut conn,
                 order.direction.opposite(),
                 OrderType::Limit,
                 false,
                 true,
             )?;
 
-            let matched_orders = match_order(&order, opposite_direction_orders)?
-                .with_context(|| format!("Could not match order {}", order.id))
-                .map_err(|e| TradingError::NoMatchFound(format!("{e:#}")))?;
+            let matched_orders = match match_order(&order, opposite_direction_orders) {
+                Ok(Some(matched_orders)) => matched_orders,
+                Ok(None) => {
+                    // todo(holzeis): Currently we still respond to the user immediately if there
+                    // has been a match or not, that's the reason why we also
+                    // have to set the order to failed here. But actually we
+                    // could keep the order until either expired or a
+                    // match has been found and then update the state correspondingly.
+
+                    orders::set_order_state(&mut conn, order.id, OrderState::Failed)?;
+                    bail!(TradingError::NoMatchFound(format!(
+                        "Could not match order {}",
+                        order.id
+                    )));
+                }
+                Err(e) => {
+                    orders::set_order_state(&mut conn, order.id, OrderState::Failed)?;
+                    bail!("Failed to match order. Error {e:#}")
+                }
+            };
 
-            tracing::info!(order_id=%order.id, trader_id=%order.trader_id, "Found a match with {} makers for new order.", matched_orders.makers_matches.len());
+            tracing::info!(trader_id=%order.trader_id, order_id=%order.id, "Found a match with {} makers for new order.", matched_orders.taker_matches.filled_with.matches.len());
 
-            let mut orders_to_set_taken = vec![matched_orders.taker_matches.filled_with.order_id];
-            let mut order_ids = matched_orders
-                .taker_matches
-                .filled_with
-                .matches
-                .iter()
-                .map(|m| m.order_id)
-                .collect();
-
-            orders_to_set_taken.append(&mut order_ids);
-            notify_traders(matched_orders, &self.authenticated_users).await?;
-
-            for order_id in orders_to_set_taken {
-                tracing::info!(%order_id, "Setting order to taken.");
-                if let Err(err) = orderbook::db::orders::set_is_taken(&mut conn, order_id, true) {
-                    let order_id = order_id.to_string();
-                    tracing::error!(order_id, "Could not set order to taken. Error: {err:#}");
-                }
+            let match_params = matched_orders.matches();
+            for match_param in match_params {
+                matches::insert(&mut conn, match_param)?;
```

Comment:
Because I want to track if a match got executed on per trader. The model would have gotten to complex otherwise.

Also considering that we could have multiple matches per order. By having a single match record we can easily reflect 1:n and m:n relations.